Feature/offline email move

blackjack2693 opened the pull request:
## What does this change?

The moving of emails to another folder is now working offline

Co-authored-by: *Johann Schulze Tast*
Co-authored-by: *christian Helms*

CLOSES #382
## Decisions / Choices I made

the implementation of a new class called ICMovedMailStorage was necessary to make sure there are no interferences with the server version while updating. the updating is now only possible if you are online, otherwise there could be inconsistencies. this is also the reason update is no longer part of the queue in CommandStorage. 

the methods to determine the highest and lowest UID in ICFolder had to be changed not to care about the moved emails, the server does not know about when being offline. otherwise the moved emails are no longer classified as moved but as normal emails of the folder they are in. 

## Checklist

- [x] The branch implementing the login changes should be merged before this one, this one based on it 
- [x] TEST IT locally to ensure there are no bugs in this version

## Note
Pull request will have to be merged manually. To prevent the disaster from last time:
- first merge `develop` into this branch (do **not** amend merge commits and be sure to find **all** merge conflicts before committing)
- then use GitHub to do the usual squash-and-merge

elenagensch committed:
test folder load offline

elenagensch committed:
Splits Login from ICFolderDialog initialization

elenagensch committed:
deletes readFoldersFromDisk

elenagensch committed:
set unitializedAccounts

elenagensch committed:
relieve passwordPromptCallback and accountsComplete

elenagensch committed:
restore multiple account login

elenagensch committed:
emails not read from disk second time in login process

elenagensch committed:
removing uninitializedAccounts instance variable

christian-helms committed:
"production ready"

christian-helms committed:
implements moving of offline emails, not tested yet

blackjack2693 committed:
still under development, further smooothing out some bugs

blackjack2693 committed:
in development, just to save progress

Commit:
Merge login into offline_email_move (#416)

copy current state of login branch

Co-authored-by: Elena Gensch <elena.gensch@student.hpi.uni-potsdam.de>
Co-authored-by: Jan Groeneveld <jan.niklas.groeneveld@gmail.com>

Commit:
now the Moving of emails should work properly offline and online, please test at your system before approving. Sould be ready to merge into develop.

laugengebaeck commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st:
```diff
@@ -6,7 +6,10 @@ checkForBadRequest
 	(ICParser responseIsError: self stream lastResponse) ifTrue: [
 		self fatalErrorOccured: true.
-		self inform: 'Wrong username or password'.
+		BreakPoint break.
+		Transcript show: 'error username'; cr.
+		self inform: 'Wrong username or passwordddddddddd'.
+		Transcript show: self stream lastResponse; cr.
```
remove debugging stuff

blackjack2693 replied:
done

laugengebaeck commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/isCommandStorageEmpty.st:
```diff
@@ -0,0 +1,4 @@
+move
+isCommandStorageEmpty
+
+	^ self storedCommands isEmptyFor: self accountInfo hash
```
we really need to find a solution for the hash method (either make it work again or use another way of identifying accounts)

blackjack2693 replied:
CH suggested using the names of the accounts, but we need to enforce it. This is currently done in the MovedMailStorage

laugengebaeck commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st:
```diff
@@ -1,14 +1,4 @@
-			self storedCommands popNextCommandFor: hash]
-		on: Error do: [conOK := false]]
\ No newline at end of file
+sendCommand: aCommand 
```
add newline :sweat_smile:

laugengebaeck commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st:
```diff
@@ -0,0 +1,2 @@
+connection
+sendNextCommands    | hash conOK |    hash := self accountInfo hash.  conOK := true.    [conOK and: (self storedCommands isEmptyFor: hash) not] whileTrue: [    [self sendDirectCommand: (self storedCommands getNextCommandFor: hash).    self storedCommands popNextCommandFor: hash]   on: Error do: [conOK := false]]
```
format method so that it isn't all in one line

Also, I noticed that ICCommandStorage has an unsqueakish method name (`getNextCommand`) in the current version on `develop` (see other comment). Maybe we could also change that somewhere.

blackjack2693 replied:
done

laugengebaeck commented on packages/IMAPClient-Core.package/ICEndPoint.class/instance/moveEmail.to..st:
```diff
@@ -1,10 +1,20 @@
+moveEmail: anEmail to: aFolderName
 	
-	self resetDeletedFlagWhenRecovering: anEmail.
+	"has to be refactored immediately"
```
maybe delete that line ^^

blackjack2693 replied:
done

laugengebaeck commented on packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st:
```diff
@@ -1,4 +1,4 @@
 
-	self emails remove: anICEmail.
\ No newline at end of file
+	self emails remove: anICEmail ifAbsent: [ Transcript show: 'ICFolder removeMail is not there' ]
```
remove transcript output

blackjack2693 replied:
done

laugengebaeck commented on packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st:
```diff
@@ -1,10 +1,16 @@
-	(self folderName = 'root') ifTrue: [^true].
 	
+	| movedEmails acc |
+	acc := self accountName.
```
where do you use acc in this method? :sweat_smile:

blackjack2693 replied:
nowhere :D will be gone

laugengebaeck commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st:
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
+deleteEmailsFromFolder: aFolder
+	"comment stating purpose of message"
```
remove line

blackjack2693 replied:
done

laugengebaeck commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st:
```diff
@@ -0,0 +1,6 @@
+as yet unclassified
+getSetForAccount: anAccountName andFolder: aFolder
```
`get` prefix in method name is "unsqueakish" (apparently)

also shares lots of code with `addMail:forAccount:` (should probably be used in that method)

blackjack2693 replied:
anyone any suggestions for a better name? :D

laugengebaeck replied:
`movedMailsForAccount:andFolder:` maybe?

blackjack2693 replied:
good idea, done

laugengebaeck reviewed (commented):
Looks good and I'm happy that this also finally works now :partying_face: However, there are some minor issues that need to be addressed.

laugengebaeck commented:
Also, please write a short class comment for the `ICMovedEmailStorage` :)

Commit:
finalizing this branch to merge.
Especially adressing comments by laugengebaeck, most small fixes without additional features or bugs.
Also adding a new dialogTitle for the case the clinets is opened but not updated.

laugengebaeck commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/throwErrorIfNotLoggedIn.st:
```diff
@@ -1,8 +1,9 @@
 	(self checkConnected) ifFalse: [
 		self fatalErrorOccured: true.
-		self inform: 'Cannot connect to the server'. 
+		self inform: 'Cannot connect to the server. Please check wether you are connected to the internet'. 
```
```suggestion
		self inform: 'Cannot connect to the server. Please check whether you are connected to the internet'. 
```

laugengebaeck commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/README.md:
```diff
@@ -0,0 +1,7 @@
+An ICMovedEmailStorage is a dictionary of dictionaries containing sets of emails.
+it is also a *singleton*.
+The first dictionary represents the account / endpoint, the second one the folder the mail belongs to. 
+The mails were typically moved to the given folder and saved here while the clinet is offline. 
```
```suggestion
The mails were typically moved to the given folder and saved here while the client is offline. 
```

laugengebaeck commented on packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st:
```diff
@@ -2,11 +2,10 @@ action - emails
 	do: [:e | self inform: e asString].
   
-	successful ifTrue: [self updateSynchTime].
+	successful ifFalse: [self updateSynchTime].
```
Why was this changed? Also, that variable is now named misleadingly.

christian-helms replied:
You're right. That's an obvious mistake.

laugengebaeck replied:
Fixed/reverted :)

Commit:
fix spelling mistakes

sm1lla commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkConnected.st:
```diff
@@ -1,8 +1,8 @@
-	(self fatalErrorOccured)
-		ifTrue: [^ false].
+	"(self fatalErrorOccured)
+		ifTrue: [^ false]."
```
Please remove comments

sm1lla commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/select.direct..st:
```diff
@@ -0,0 +1,6 @@
+fetch
+select: aFolderName direct: aBool
```
Maybe call this method `select: aFolderName directly: aBool` instead. I find it a bit confusing at the moment that it gets a boolean as a parameter.

elenagensch replied:
I agree

sm1lla commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st:
```diff
@@ -0,0 +1,10 @@
+connection
+sendNextCommands   
+	
+	| hash conOK |    
```
Please rename `conOK` to `isConnected` or something similar

sm1lla commented on packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st:
```diff
@@ -2,6 +2,5 @@ login
 	"self connectAndLoginWith: anAccountInfo.
- 
-	^ self fatalErrorOccured not"
-	^ true
```
Please remove comments

sm1lla commented on packages/IMAPClient-Core.package/ICFolder.class/instance/highestUniqueID.st:
```diff
@@ -1,6 +1,7 @@
-	^ (self emails detectMax: [:mail | mail uniqueID asNumber]) uniqueID
\ No newline at end of file
+	movedEmailSet := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	^ (self emails detectMax: [:mail | (movedEmailSet includes: mail) ifFalse: [mail uniqueID asNumber]ifTrue: [-1]]) uniqueID
```
-1 is a magic number

sm1lla commented on packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st:
```diff
@@ -1,6 +1,7 @@
-	^ (self emails detectMin: [:mail | mail uniqueID asNumber]) uniqueID
\ No newline at end of file
+	movedEmailSet := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	^ (self emails detectMin: [:mail | (movedEmailSet includes: mail) ifFalse: [mail uniqueID asNumber] ifTrue: [9999999]] ) uniqueID
```
999999 is a magic number

christian-helms replied:
Yeah, that's weird.

blackjack2693 replied:
it makes sure that while fetching new mails the moved ones, which could have the same UID as a mail already in the system, the added email is ignored.

sm1lla commented on packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st:
```diff
@@ -1,4 +1,4 @@
 
-	self emails remove: anICEmail.
\ No newline at end of file
+	self emails remove: anEmail ifAbsent: []
```
what does ifAbsent: [] do?

christian-helms replied:
Indeed it is unnecessary as OrderedCollection inherits from Collection which implements an element-based remove with a default specification for the case that the element is not found.

sm1lla commented on packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st:
```diff
@@ -1,10 +1,15 @@
 	
+	| movedEmails |
+	movedEmails := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	(self folderName = 'root') ifTrue: [ ^true].
```
`[^ true]`

sm1lla commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st:
```diff
@@ -0,0 +1,6 @@
+accessing
+movedMailsForAccount: anAccountName andFolder: aFolder
+	
+	 | accountDictionary |
```
new line

sm1lla commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st:
```diff
@@ -0,0 +1,6 @@
+	
+	 | accountDictionary |
+	accountDictionary := self at: anAccountName ifAbsent: [self at: anAccountName put: Dictionary new].
+	^(accountDictionary at:  aFolder ifAbsent: [ accountDictionary at: aFolder put: Set new.]) 
```
Space after `^`

sm1lla commented on packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st:
```diff
@@ -3,8 +3,8 @@ switchToNewEndpoint
 	self selectedEndpoint ifNil: [^self].	
-	self folderWith: self selectedEndpoint.
-	
+	"self folderWith: self selectedEndpoint."
```
Please remove comment

sm1lla reviewed (approved):
Works nicely! Just some comments on the style.

sm1lla reviewed (changes requested):
But please address the failing tests.

elenagensch commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st:
```diff
@@ -6,7 +6,7 @@ checkForBadRequest
 	(ICParser responseIsError: self stream lastResponse) ifTrue: [
 		self fatalErrorOccured: true.
-		self inform: 'Wrong username or password'.
+		self inform: 'responseError in checkForBadRequest'.
```
We should keep in mind that it might be confusing for the users when they login with wrong username or password and get this message (instead of wrong username or password). Error messages for users should be thrown in higher layers (when you know what operation failed)

Clearer error messages could be a good first issue for the next group.

blackjack2693 replied:
good idea

elenagensch commented on packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st:
```diff
@@ -0,0 +1,9 @@
+	accountName := aFolder accountName.
+	emailSet := self movedMailsForAccount: accountName andFolder: aFolder.
+	emailSet do: [:anEmail | aFolder endpoint deleteMailLocally: anEmail].
+	
```
newline

elenagensch commented on packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st:
```diff
@@ -0,0 +1,10 @@
+		anEndpoint accountInfo password: (Project uiManager requestPassword: 'Please enter your password!').
+		anEndpoint accountInfo password ifEmpty: [^false].
+		continueAsking := (self passwordPromtCallback: anEndpoint) not].
+	^ true.
```
remove point at the end

elenagensch commented on packages/IMAPClient-UI.package/ICFolderDialog.class/instance/passwordPromtCallback..st:
```diff
@@ -1,13 +1,9 @@
-		ifTrue: [self accountsComplete]
-		ifFalse: [ICPasswordDialog openWith: (self uninitializedAccounts first) and: self]
\ No newline at end of file
+	^true
```
^ true

christian-helms commented on packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st:
```diff
@@ -1,14 +1,5 @@
-	
-	self storedCommands add: aCommand for: self accountInfo hash.
-	
-	[conOK and: (self storedCommands isEmptyFor: hash) not] whileTrue: [
```
We should start identifying accounts by their name or a dedicated identifier instead of some weird hash.

blackjack2693 replied:
i am now using the email address, this should be unique 😉

laugengebaeck committed:
resolve merge conflicts

laugengebaeck committed:
address review comments

blackjack2693 committed:
The commit fixes the briken tests and addresses some minor problems. 
The tests required a variable not provided by the environment
